tools/ci_build/rewriter.py

```diff
@@ -0,0 +1,226 @@
+
+        for operator in graph.node:
+
+            if operator.op_type not in operators:
+                operators[operator.op_type] = {}
+
+            for opset in opsets:
+                mapped_domain = map_domain(opset.domain)
+                if mapped_domain not in operators[operator.op_type]:
+                    operators[operator.op_type][mapped_domain] = []
+                if opset.version not in operators[operator.op_type][mapped_domain]:
+                    operators[operator.op_type][mapped_domain].append(opset.version)
+
+            for attr in operator.attribute:
+
+                if attr.type == AP.GRAPH: #process subgraph
+                    extract_ops_from_graph(attr.g, opsets, operators)
+
+                elif attr.type == AP.GRAPHS: #process all subgraphs
+                    for subgraph in attr.graphs:
+                        extract_ops_from_graph(subgraph, opsets, operators)
+
+    for root, _, files in os.walk(model_path):
+        for file in files:
+            if file.endswith('.onnx'):
+                model_path = os.path.join(root, file)
+                model = onnx.load(model_path)
+                extract_ops_from_graph(model.graph, model.opset_import, referred_ops)
+
+    return referred_ops #end of extract_ops_from_model(...)
+
+
+def rewrite_provider(model_path, file_path, ep_path):
+    '''rewrite provider file to exclude unused ops'''
+
+    onnx_op = 'ONNX_OPERATOR_KERNEL_CLASS_NAME'
+    onnx_op_len = len(onnx_op)
+    onnx_typed_op = 'ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME'
+    onnx_typed_op_len = len(onnx_typed_op)
+    onnx_versioned_op = 'ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME'
+    onnx_versioned_op_len = len(onnx_versioned_op)
+    onnx_versioned_typed_op = 'ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME'
+    onnx_versioned_typed_op_len = len(onnx_versioned_typed_op)
+    version_map = {} #{op:{domain:[opset1, opset2, opset3 ...]}
+    operators = extract_ops_from_file(file_path, extract_ops_from_model(model_path, {}))
+
+    def fill_version_map(op_type, opset_from, opset_to, domain):
+        '''callback func to register op in version_map'''
+
+        if op_type in version_map:
+            if domain in version_map[op_type]:
+
+                if opset_from not in version_map[op_type][domain]:
+                    version_map[op_type][domain].append(opset_from)
+
+                if opset_to not in version_map[op_type][domain]:
+                    version_map[op_type][domain].append(opset_to)
+            else:
+                version_map[op_type][domain] =\
+                    [opset_from, opset_to] if opset_from < opset_to else [opset_from]
+
+        else:
+            version_map[op_type] =\
+                {domain: [opset_from, opset_to] if opset_from < opset_to else [opset_from]}
+
+        return True #end of fill_version_map(...)
+
+
+    def need_comment(op_type, opset_from, opset_to, domain):
+        '''callback func to check if the op is in ops'''
+
```

Comment:
nit: would 'disable_op' be a clearer name?

Reply:
Yeah why not.